Add doorsense sensor for August 3rd Gen Smart Lock Pro

ehendrix23 opened the pull request:
## Description:

The August 3rd Gen Smart Lock Pro comes with a DoorSense sensor to identify if the door is open or closed.

I added a new binary sensor within the august component to expose the sensor to HASS. A binary sensor will be created for each lock that has the doorsense sensor available as well. During initialization it checks if the status returned is not Unknown and only if it is Unknown it will add the device, otherwise the binary sensor will not be added.

I further also added debug logging in various areas of the coding for both locks and doorbells.

This is a re-do from PR 17116 https://github.com/home-assistant/home-assistant/pull/17116 that I closed due to rebase issue on my end.

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6488

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

  - [X] New dependencies are only imported inside functions that use them ([example][ex-import]).

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

ehendrix23 committed:
Add doorsense sensor for August 3rd Gen Smart Lock Pro

Add a binary sensor to August for the August 3rd Gen Smart Lock Pro doorsense sensor.

This is a re-do from PR 17116 https://github.com/home-assistant/home-assistant/pull/17116 that I closed due to rebase issue on my end.

2016for commented:
Awesome. Was waiting for this for quite some time. Tried doing it myself but couldn't. :)
Is there a way I can try it out on my current HA and report back? Will that help in getting this approved or something?
Very new to this stuff. :)
Thanks for your time and effort.

2016for commented:
Plus I am hopeful the Logger will help me identify the excessive amount of delay (>30-45 seconds) I get right now when someone presses the bell/get status of lock/control the lock.

balloob commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -60,24 +68,55 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         """Return true if the binary sensor is on."""
+        from august.lock import LockDoorStatus
+
+        # For door sensor, return true if open or unknown,
```
If status is unknown, we should return `False` from the `available` property.

balloob replied:
I think you misread my comment. You need to return `False` from the **available** property.

ehendrix23 replied:
@balloob ,
Found that snjoetw had provided updates for this component but those were never introduced into HASS. 
Just took his source (as it seemed cleaner and better to me) and added the debug logging I had done; not creating the binary_sensor if the DoorSense sensor does not exist; and fix for issue when battery attribute is not available.

balloob commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -60,24 +68,55 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                continue
+
+            state_provider = SENSOR_TYPES[sensor_type][2]
+            if state_provider(data, lock) is LockDoorStatus.UNKNOWN:
```
We shouldn't filter out anything, that means that they might just be unavailable during boot and will then require a restart once fixed up by the user. Instead we should return `False` from the available property when status is unknown and still add them to HA.

balloob replied:
I don't understand how you can have handling in is_on for unknown which is for real devices, and then filter out unknowns that are not real devices. How can you distinguish which is which?

Can't we look at the model and have a flag if they enable sensors?

ehendrix23 replied:
@balloob, 

There does not seem to be a way to really determine if a sensor is there. We can get the model number but that does not automatically mean there is a doorsense sensor as that is added separately as part of the installation for this lock.
Further, if we were to look at the actual model number then we would need to keep on updating based on new models they might provide.

ehendrix23 committed:
Changed to use snjoetw provided code

Going through the py-august I found that snjoetw had provided updated versions for the august component (august.py and binary_sensor/august.py) to include DoorSense sensor.
Changed what I did to to what snjoetw provided instead as he split it into 2 classes; much cleaner I think.

I modified his coding with:
   Fixes that were done to the August component and not part of the coding snjoetw provided.
   Added the debug logging improvement I had done in the code.

Note, fix I committed earlier for lock atribute (lock/august.py) is thus still the same.

ehendrix23 committed:
Reverted change from add_device to add_entities

Missed an item when merging snjoetw's code with current. Fixed.

ehendrix23 commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -46,26 +56,84 @@ def _activity_time_based_state(data, doorbell, activity_types):
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+def setup_platform(hass, config, add_devices, discovery_info=None):
```
Apologies, missed this one. Fixed.

ehendrix23 committed:
Updated call from add_devices to add_entities as well

Updated the call from add_devices to add_entities.

MartinHjelmare reviewed (commented):
Please revert the access mode changes to the files.

ehendrix23 commented:
@MartinHjelmare Not sure what you mean with "revert access mode changes to the files". 
When I check in my repository using "git ls-files -s august.py" then I see 10644; same as I do for other files in the repository I have not touched. 

Appreciate your guidance.

MartinHjelmare commented:
Go to the changes tab in this PR. You can see it says:
```
homeassistant/components/august.py 100644 → 100755
```
It might have to do with what `git` client you use.

ehendrix23 committed:
Fixed permissions on files

Fixed permissions on components/august.py and binary_snesor/august.py

ehendrix23 commented:
@MartinHjelmare Yep, now I saw it. Went back to my local and realized it was set to a different branch when I checked it. Set it to the correct branch and then I could also see those files were set to 755. Changed to 644 and did a commit. Now it is not stating anymore the permissions have changed.

Thanks for the pointer!

MartinHjelmare commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -58,14 +72,68 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+                    "Not adding sensor class %s for lock %s ",
+                    SENSOR_TYPES_DOOR[sensor_type][1], door.device_name
+                )
+            else:
```
We could replace this with `continue`. Then we save an indentation below.

ehendrix23 committed:
Changed if/else to if/continue

Changed logic so that if the door sensor state is unknown during initalization the debug log is written and then continue the loop instead of using if/else logic.

balloob commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -58,14 +72,69 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
+    def update(self):
+        """Get the latest state of the sensor."""
+        state_provider = SENSOR_TYPES_DOOR[self._sensor_type][2]
+        self._state = state_provider(self._data, self._door)
```
This does not return a boolean, however it's returned directly from is_on. That is wrong.

balloob replied:
As I have written twice now, and this is the third time, you need to return `False` from the available property if the status is unknown.

ehendrix23 replied:
@balloob,

My sincere apologies, but I thought that this coding was doing what you were asking for returning False if the status is unknown.

For
```
state_provider = SENSOR_TYPES_DOOR[self._sensor_type][2]
```
sets state_provider to method _retrieve_door_state.

Thus
`self._state = state_provider(self._data, self._door)`
for the door sensor would be:
`self._state = _retrieve_door_state(data, lock)`

Method _retrieve_door_state(data, lock) returns:
`return doorstate == LockDoorStatus.OPEN`

Thus self._state would be set to True (if door is open) or False (if door state is anything else).

Is this not what you mean? I looked at some other code and for example in spc.py I see for is_on the return is:
`return self._zone.input == ZoneInput.OPEN`

Thus I think that the current coding sets the state indeed then to True or False with only setting it to True if the returned status is Open, otherwise returning False. 

If this is incorrect then my sincere apologies and I would ask if you could point me to some coding of a sensor that does it correctly. I looked at a few and they all return True or False bu doing it directly in is_on or by setting self._state somewhere else to True or False and then just returning self._state

Thank you.

ehendrix23 replied:
@balloob, thanks for pointing me in the right direction for the available property. It is defined now for the doorsensor and set to False if self._state is unknown otherwise it is set to True.

balloob commented:
Your responses to my comments do not make any sense. Instead of fixing the issues I point out, you claim you took snoets code and just put it up as a PR? That's not a reason for me to accept bad code. And we won't. 

Your suggestion that we should interpreted sensors that are unknown as non-existing, but later interpreted unknown as not available makes 0 sense. Something means one or another.

ehendrix23 commented:
@balloob,

Reviewing snoets code, I  thought that it also covered the items you had mentioned and I liked his approach better with making a different class for doorbell and another one for door as it seemed cleaner to me. 

I provided in the PR comments for py-august that I used what he provided in this PR to add it to HASS, ensuring he receives the credit and that I am not claiming his work for my own. I did mention in the comments the changes I did to the code he provided.
Looking back at the comment, the 1 item I missed in there (and I apologize for that) is that I included the check if the door status is unknown before adding the sensor.

I can easily remove this however then that does mean that anyone with the August Smart lock will get a binary sensor in HASS irrespective if they have the door lock with the sensor or not. My thoughts were that HASS should not be creating entities that are non-existent and hence I updated the code to filter out any locks that do not have the sensor.

The status unknown would only be there if py-august did not get Open or Closed for the doorstate sensor. Hence if the lock does not have the sensor then status unknown would be returned and that is why I'm using that status to filter out any doorlocks that do not have this sensor.

I was wrong on the status unknown being returned if for some reason py-august is unable to retrieve it. Instead an exception is raised within py-august if the HTTP return code is <> 200.

ehendrix23 committed:
Added available property for Door Sensor

Added the available property for the Door Sensor and setting it to False if a status unknown is received.

MartinHjelmare commented on homeassistant/components/binary_sensor/august.py:
```diff
@@ -133,6 +139,10 @@ def update(self):
         self._state = state_provider(self._data, self._door)
 
+        from august.lock import LockDoorStatus
+        self._available = False if self._state == LockDoorStatus.UNKNOWN\
```
```py
self._available = self._state != LockDoorStatus.UNKNOWN
```

ehendrix23 replied:
Updated locally. Learning every second right now about Python. :-)

Going to wait till balloob has been able to review before doing commit to see if he has any further comments.

balloob replied:
that's fine 👍

ehendrix23 committed:
Updated setting self._available

Changed line for setting self._available to what Martin provided. Much  more efficient to read. :-)

2016for commented:
@balloob @ehendrix23 It seems like new changes have been made. Let me know if there's anything we can do to get the ball rolling.

Thanks
